Prevent re-review and dismiss review actions on closed and merged PRs

kemzeb opened the pull request:
Resolves #29965.

---
Manually tested this by:
- Following the [installation](https://docs.gitea.com/next/installation/install-with-docker#basics) guide (but built a local Docker image instead)
- Creating 2 users, one who is the `Owner` of a newly-created repository and the other a `Collaborator`
- Had the `Collaborator` create a PR that the `Owner` reviews
- `Collaborator` resolves conversation and `Owner` merges PR

And with this change we see that we can no longer see re-request review button for the `Owner`:

<img width="1351" alt="Screenshot 2024-03-25 at 12 39 18 AM" src="https://github.com/go-gitea/gitea/assets/60799661/bcd9c579-3cf7-474f-a51e-b436fe1a39a4">

lunny commented:
Maybe closed pull requests cannot be reviewed anymore.

kemzeb commented:
> Maybe closed pull requests cannot be reviewed anymore.

Thanks for the quick response! I actually never seen such an edge case in the wild before.

I also just realized that in the issue it explains they can still re-request or delete a review request when the PR has already been merged. So if I understand this correctly, this looks like more than just hiding a button when a PR is closed/merged; rather, some backend logic needs to be changed so that users cannot re-request a review on a merged PR. Still kind of fuzzy how closed PRs should be handled, but I'll make this a draft PR and see if I can resolve this.

kemzeb commented:
I believe I found the culprit route handler, that being `UpdatePullReviewRequest()`. I manually tested the code path by doing what I did in my original description (this time running Gitea in a dev environment rather than having to fire up the containerized app). I got the following logs when trying to re-request a review of a merged PR:
```console
2024/03/25 10:43:00 ...rs/web/repo/issue.go:2417:UpdatePullReviewRequest() [W] UpdatePullReviewRequest: cannot add or remove a review for merged PR gek/test-repo#3
2024/03/25 10:43:00 ...eb/routing/logger.go:102:func1() [I] router: completed POST /gek/test-repo/issues/request_review for 127.0.0.1:56852, 403 Forbidden in 32.2ms @ repo/issue.go:2387(repo.UpdatePullReviewRequest)
```

I'll make this PR ready-to-review. I do have some questions/comments:
- I did not see any tests for `UpdatePullReviewRequest()`. Should I add one to account for this bug and if so should I add it to issue_test.go?
- Is there an api/v1 route I should look out for too? Maybe /api/v1/repo/pull_review.go?
- Closed PRs were discussed here. I believe closed PRs can still be reviewed since the HTML template if check I add was just to check if the PR was merged, not if it was closed (which I think we can do `$.Issue.IsClosed` for that). Then again I am a complete newbie so I may be wrong.

Thanks in advance for your time!

silverwind commented:
I checked on GitHub, it's not possible to submit a review it either closed or merged states.

BTW, we don't have to blow up this PR with more scope extensions, I could accept it as-is.

silverwind commented:
I tested changing reviews and GitHub does allow adding/removing review requests in both merged and closed states. Only the "re-review" and "dismiss" options are not possible in those states in GitHub.

kemzeb commented:
> I tested changing reviewers and GitHub does allow adding/removing review requests in both merged and closed states. Only the "re-review" and "dismiss" actions are not possible in those states in GitHub.

Thanks for the review! Reading this, it looks like simply throwing a `http.StatusForbidden` when requesting for a review on a merged PR is not good enough. I guess I need to figure out if we were given a re-review or dismiss action for the reviewer. Looking so far, I'm unsure if this is something the frontend communicates but I'll see if I can find a solution 👍

wxiaoguang commented on routers/web/repo/issue.go:
```diff
@@ -2413,6 +2413,15 @@ func UpdatePullReviewRequest(ctx *context.Context) {
 			return
 		}
+		if issue.PullRequest.HasMerged {
+			log.Warn(
+				"UpdatePullReviewRequest: cannot add or remove a review for merged PR %-v#%d",
+				issue.Repo.FullName(), issue.PullRequest.Index,
+			)
```
Overall LGTM, I think it doesn't need to output any log, since it is an expected behavior.

As a quick fix, I could vote my approval.

To have a complete solution, I have some questions:

1. Is it possible to add some tests?
2. Is there anything need to change for related API? (if some related API endpoints exist)

kemzeb replied:
Thanks for the review @wxiaoguang! I have been working on this throughout the day and think my most recent changes handles the "re-review" and "dismiss" actions, though the quality is questionable (forgive me rusty Go code!).

> Is it possible to add some tests?

I think this would be good too. I'll take a look at how unit testing works here and see if I can make some unit tests.

> Is there anything need to change for related API? (if some related API endpoints exist)

This was something I was thinking about too. When I get the time, I'll see what I can do.

Also, with my current changes, it doesn't account for teams, so I believe they will have the same weird behavior (haven't checked yet). So I might need to handle this in this PR too

kemzeb replied:
I also did not consider handle closed PRs, but it's up to you if you want me to tackle that too

wxiaoguang replied:
It's better to handle "closed PRs" together, since the new code seems well structured.

Maybe `IsErrReReviewRequestOnMergedPR` -> `IsErrReReviewOnClosedPR` ? IMO "merged" also means "closed"

kemzeb commented:
Most recent change also considers closed PRs when a user is trying to perform a re-review or dismiss request.

Will look into how unit testing works here and how I can cover these code paths.

kemzeb commented on models/issues/review.go:
```diff
@@ -618,9 +635,33 @@ func AddReviewRequest(ctx context.Context, issue *Issue, reviewer, doer *user_mo
+			return nil, err
+		}
+
+		if err := issue.LoadPullRequest(ctx); err != nil {
+			return nil, err
+		}
+		if issue.PullRequest.HasMerged {
```
Just want to make sure and ask is it safe to assume that the issue is a PR or should I squeeze in a sanity check here somewhere?

lunny replied:
The check should be on the top of this function whether the `issue.IsPull`.

kemzeb replied:
I couldn't find this check above my changes (or within the called functions above them), so I decided to add this check as closed to my code as possible.

Is this alright or should I instead move this check to the top and return an error if it's not true (I guess it acts like an assertion here to make sure we're only calling these functions when dealing with PRs)

lunny commented on models/issues/review.go:
```diff
@@ -618,9 +635,33 @@ func AddReviewRequest(ctx context.Context, issue *Issue, reviewer, doer *user_mo
+		}
+
+		if issue.IsClosed {
+			var err error = ErrReviewRequestOnClosedPR{}
```
It's very weird. It will succeed but users will get a failure information?

kemzeb replied:
Hmm, I'm not quite sure what you mean by succeeds. To me I see this as:

- The user is trying to perform an add review request
- The reviewer has already made a review before (which means to me that this is a re-review request)
- The PR is closed, so we should return the error `ErrReviewRequestOnClosedPR{}` to callers

But now that I'm reading the code again I think I might see what you mean. Does the problem have to do with the transaction being committed that I do when I return a error?

lunny replied:
I mean we should not Commit the transaction, just let it rollback.

kemzeb replied:
Gotcha. Being this low I wasn't too sure on how to handle database transactions and wrongly assumed from reading the existing comment here that I should commit no matter if it was a success or failure. Will make the necessary changes.

kemzeb replied:
Fixed

kemzeb commented:
My recent commit does the following:
- Rollback failures in `review.AddReviewRequest()` instead of committing them
- Throw the PR merge checking logic under a `if issue.IsPull` branch
- Fix a UI logic bug
- Fix error messages to include both  **closed** and merged

kemzeb commented:
I have added a couple of unit tests to cover the code paths I've added. Let me know if I should add more or make improvements.

Next I'll scour the API routes looking for any signs of issues similar to this one.

kemzeb commented:
(Sorry for the repeated force-pushes)

kemzeb commented:
Made the necessary changes to the API and from testing using the swagger API client it looks like it works as expected. I haven't seen any tests related to the API code, but let me know if I should add some for my changes.
